amaranth/build/plat.py

```diff
@@ -315,6 +315,45 @@ def toolchain_prepare(self, fragment, name, **kwargs):
 
         rtlil_text, self._name_map = rtlil.convert_fragment(fragment, name=name)
 
+        # Retrieve an override specified in either the environment or as a kwarg.
+        # expected_type parameter is used to assert the type of kwargs, passing `None` will disable
+        # type checking.
+        def _extract_override(var, *, expected_type):
+            deprecated_var_env = "NMIGEN_{}".format(var)
+            var_env = "AMARANTH_{}".format(var)
+            if deprecated_var_env in os.environ or var_env in os.environ:
+                # On Windows, there is no way to define an "empty but set" variable; it is tempting
+                # to use a quoted empty string, but it doesn't do what one would expect. Recognize
+                # this as a useful pattern anyway, and treat `set VAR=""` on Windows the same way
+                # `export VAR=` is treated on Linux.
+                if var_env in os.environ:
+                    var_env_value = os.environ[var_env]
+                elif deprecated_var_env in os.environ:
+                    var_env_value = os.environ[deprecated_var_env]
+                return re.sub(r'^\"\"$', "", var_env_value)
+            elif var in kwargs:
+                if expected_type is None or isinstance(kwargs[var], expected_type):
+                    return kwargs[var]
+                else:
+                    raise TypeError("Invalid keyword argument '{}' of type '{}' when type '{}' was expected."
+                                    .format(var, type(kwargs[var]).__name__, expected_type.__name__))
+            else:
+                return jinja2.Undefined(name=var)
+
+        def get_override(var):
+            value = _extract_override(var, expected_type=str)
+            return value
+
+        def get_override_flag(var):
+            value = _extract_override(var, expected_type=bool)
+            if isinstance(value, str):
+                value = value.lower()
+                if value in ("0", "no", "n", "false", ""):
+                    return False
+                if value in ("1", "yes", "y", "true"):
+                    return True
```

Comment:
I think unrecognized values shouldn't be silently swallowed here (and converted to `True` by the fallthrough below).